New logic for default parameters in IA classes

jngrad opened the pull request:
Fixes #3069

Change internals of bonded/non-bonded IA Python classes:
- `valid_keys()` and `required_keys()` now return sets instead of tuples/lists/strings (strings can lead to cryptic error messages since they are iterable)
- `default_params()` and `set_default_params()` now return values only for optional parameters
- wrong keys (typos or unknown keys) in optional parameters were fixed
- added tests for a few common IAs to check the new syntax and verify default parameters have the correct name and don't overlap with required parameters, which can't have default values

jngrad committed:
Make valid_keys/required_keys sets

Avoids the situation where a string is returned (which is iterable).

jngrad committed:
Prevent override of min/max in Tabulated classes

jngrad committed:
Return only optional values in default_params()

jngrad committed:
Add tests for new bonded/nonbonded IA requirements

Check the new Python interface for bonded/nonbonded IAs:
- valid_keys() and required_keys() return sets
- default_params() and set_default_params() return values only for
  optional parameters

KaiSzuttor commented on src/python/espressomd/interactions.pyx:
```diff
@@ -487,18 +479,7 @@ IF LENNARD_JONES_GENERIC == 1:
-                "b2": 0.,
-                "delta": 0.,
-                "lam": 0.}
+            return {"delta": 0., "lam": 0.}
```
why are only those two parameters default ones? Shouldn't we reflect the facts from the core here?

jngrad replied:
* `lam` is now 1 by default to reflect the core
* this line is now guarded by `LJGEN_SOFTCORE`
* keeping them as optional allows this generic LJ class to behave the same way when `LJGEN_SOFTCORE` is enabled, unless the user provides custom values for `lam` and `delta`

jngrad committed:
Fix value mismatch in generic LJ Softcore

Use same default value of 1 for lambda in both the core and Python
interface. Replace the silent range check in the core by a verbose
range check in the Python interface.

jngrad committed:
Formatting

jngrad committed:
Fix typo

jngrad commented:
@KaiSzuttor is anything missing?

KaiSzuttor reviewed (approved):
LGTM in the future we probably want to have constructors for the interactions in the core. Or at least on the python level.

jngrad commented:
Agreed. The Python class should only take care of checking variable types, the core should take care of checking values. Core constructors are planned in #3037.

jngrad commented:
bors r=KaiSzuttor

jngrad commented:
bors cancel

jngrad commented:
bors r-

jngrad commented:
bors r=KaiSzuttor
